### PR TITLE
Enable using dash or underscore for CLI args

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

Alternative approach to https://github.com/huggingface/accelerate/pull/2525

This PR adds a pinch of magic for users that are used to doing `--a-b` with arguments that pass-through to things such as `torchrun`, accelerate can process those if we have them defined ourselves already.

E.g. a user can do:
```bash
accelerate launch --no_python --multi_gpu --num_processes 8 --monitor_interval=1 ./trampoline.sh python3 -c "print('hello')"
```
**OR**:
```bash
accelerate launch --no-python --multi-gpu --num-processes 8 --monitor-interval=1 ./trampoline.sh python3 -c "print('hello')"
```

Fulfills some of https://github.com/huggingface/accelerate/issues/2241

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan

### CODE REVIEW

**Review by BenjaminBossan — APPROVED** [review]
This PR looks more robust to me than the previous one and almost completely avoid private methods, nice.

I left a few comments, but those are only nits.

File: tests/test_cli.py [hunk 1]
```diff
@@ -99,6 +100,50 @@ def test_notebook_launcher(self):
...
+        assert result.multi_gpu is True
+        assert result.num_processes == 4
+        # And use a mix
+        args = ["--multi_gpu", "--use_deepspeed", "--use_fsdp", "--num-processes", "4", "test.py"]
+        result = parser.parse_args(args)
+        assert result.multi_gpu is True
+        assert result.use_deepspeed is True
+        assert result.use_fsdp is True
+        assert result.num_processes == 4
+
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply tests/test_cli.py:145]: Looks like we can use `print_help`, I'll see if I can get a test for that going

File: src/accelerate/commands/launch.py [hunk 2]
```diff
@@ -123,20 +124,32 @@ def __call__(self, parser, namespace, values, option_string=None):
                 if all([arg.help == argparse.SUPPRESS for arg in group._group_actions]):
                     parser._action_groups.remove(group)
 
+        for i, opt in enumerate(opts):
+            new_strings = []
+            for option in opt.option_strings:
+                if "-" not in option[2:]:
+                    new_strings.append(option)
+            opts[i].option_strings = new_strings
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/commands/launch.py]: Done, we can actually simplify this more now with the new method:

```python
action.option_strings = [s for s in action.option_strings if "-" not in s[2:]]
```

File: src/accelerate/commands/utils.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    def __init__(self, option_strings, dest, const, default=None, required=False, help=None, metavar=None):
+        super().__init__(
+            option_strings=option_strings,
+            dest=dest,
+            nargs=0,
+            const=const,
+            default=default,
+            required=required,
+            help=help,
+        )
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/commands/utils.py:51]: Updated string, however we can't do `*args, nargs=0, **kwargs` I found

File: src/accelerate/commands/utils.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        self,
+        option_strings,
+        dest,
+        default=None,
+        required=False,
+        help=None,
+    ):
+        super().__init__(
+            option_strings=option_strings, dest=dest, const=True, default=default, required=required, help=help
+        )
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/commands/utils.py:72]: Nope as argparse doesn't read them in like that/applies them in that way which makes it very messy and annoying.

File: src/accelerate/commands/utils.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+            option_strings=option_strings,
+            dest=dest,
+            nargs=0,
+            const=const,
+            default=default,
+            required=required,
+            help=help,
+        )
+
+    def __call__(self, parser, namespace, values, option_string=None):
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/commands/utils.py:53]: Was needed, forgot we need to pass the const 😓

**Review by BenjaminBossan — APPROVED** [review]
Nice work, well done, this looks cleanest from all the implementations. I also did a double check by running a diff on `accelerate launch --help` and, apart from the added sentence ("Launch a python script..."), the two are identical.

I had one question, but it's no blocker for merging.

File: src/accelerate/commands/launch.py [hunk 6]
```diff
@@ -63,80 +64,93 @@
...
+        self.titles = [
             "Hardware Selection Arguments",
             "Resource Selection Arguments",
             "Training Paradigm Arguments",
             "positional arguments",
             "optional arguments",
         ]
+
+    def add_argument(self, action: argparse.Action):
+        if "accelerate" in sys.argv[0] and "launch" in sys.argv[1:]:
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply src/accelerate/commands/launch.py:101]: You can run this command as either `accelerate launch` or `accelerate-launch`. This catches the former.

[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply src/accelerate/commands/launch.py:101]: Oh, I didn't know about `accelerate-launch`.

**Pull request merged**